src/create/schema/index.ts

```diff
@@ -83,132 +83,145 @@ export const createSchema = <
   zodSchema: ZodType<Output, Def, Input>,
   state: SchemaState,
 ): oas31.SchemaObject | oas31.ReferenceObject => {
-  if (zodSchema._def.openapi?.type) {
-    return createManualTypeSchema(zodSchema);
-  }
+  if (state.visited?.has(zodSchema)) {
+    throw new Error(
+      `The schema ${JSON.stringify(
+        zodSchema._def,
+      )} needs to be registered because it's circularly referenced`,
+    );
+  }
+  state.visited ??= new Set();
+  state.visited.add(zodSchema);
+  try {
```

Comment:
@samchungy actually maybe I'll replace this `try/finally` with a helper function

Reply:
seems odd that it would, let me investigate...

Reply:
hmmm it does, but I know for a fact that not removing nodes from a visited set can cause problems in cycle detection algorithms...

Reply:
If you figure out how to break it let me know. It's almost 1am over here so I'm gonna bugger off to bed so let me know if you can break it 😄 

https://github.com/samchungy/zod-openapi/releases/tag/v2.4.0-beta.1

Reply:
okay I figured out how to break it; `state.lazy` wasn't getting shared between the two properties because it wasn't initialized at that point, but if I wrap `UserSchema` in a `lazy`, `state.lazy` does get shared between the two properties and triggers the bug

Reply:
I just added code to print the path to the circular reference so that it's easier for users to find the problem!

```ts
  it('throws errors when cycles without refs are detected', () => {
    const cycle1: any = z.lazy(() => z.array(z.object({ foo: cycle1 })));
    expect(() => createSchemaOrRef(cycle1, createOutputState())).toThrow(
      `The schema at lazy schema > array items > property: foo needs to be registered because it's circularly referenced`,
    );
    const cycle2: any = z.lazy(() => z.union([z.number(), z.array(cycle2)]));
    expect(() => createSchemaOrRef(cycle2, createOutputState())).toThrow(
      `The schema at lazy schema > union option 1 > array items needs to be registered because it's circularly referenced`,
    );
    const cycle3: any = z.lazy(() => z.record(z.tuple([cycle3.optional()])));
    expect(() => createSchemaOrRef(cycle3, createOutputState())).toThrow(
      `The schema at lazy schema > record value > tuple item 0 > optional needs to be registered because it's circularly referenced`,
    );
  });
```